src/main/scala/zio/interop/reactiveStreams/Adapters.scala

```diff
@@ -46,10 +40,145 @@ object Adapters {
     } yield (error, demandUnfoldSink(subscriber, demand))
 
   def publisherToStream[A](publisher: Publisher[A], bufferSize: Int): ZStream[Any, Throwable, A] =
-    Stream.unwrap(
-      QueueSubscriber.make[A](bufferSize).flatMap {
-        case (subscriber, stream) => UIO(publisher.subscribe(subscriber)).const(stream)
-      }
+    new ZStream(
+      for {
+        (q, subscription, completion, subscriber) <- makeSubscriber[A](bufferSize).toManaged_
+        _                                         <- UIO(publisher.subscribe(subscriber)).toManaged_
+        sub <- subscription.await.interruptible
+                .toManaged(sub => UIO { sub.cancel() }.whenM(completion.isDone.map(!_)))
+        process <- process(q, sub, completion)
+      } yield process
     )
 
+  def sinkToSubscriber[R, R1 <: R, A1, A, B](
+    sink: ZSink[R, Throwable, A1, A, B],
+    bufferSize: Int
+  )(subscribe: Subscriber[A] => ZIO[R1, Throwable, Unit]): ZIO[R1, Throwable, B] =
```

Comment:
Here's a suggestion:
- Return a `ZManaged[(Subscriber[A], ZIO[R1, Throwable, B])]`. Yes, it's ugly, but I don't think you can avoid that
- The acquisition for the managed is the `makeSubscriber` part, with appropriate finalizers
- The rest of this function gets stuffed into the `ZIO[R1, Throwable, B]`. When you run that, you create the stream and run it into the sink.

Does that make sense?

Reply:
Wouldn't that allow to create multiple Streams from the Subscriber (which can't work) by running the ZIO multiple times?

Reply:
Right. You could make that memoized inside the scope of the `ZManaged`, or make that fail after the second time. Or just document it :-)

Reply:
Memoization does not work, as we can't memoize a Stream. Failing the second time, even when documented, does not feel better than a continuation to me. The continuation might not be beautiful, but it allows for exactly the intended interaction and makes this kind of error impossible.

Reply:
Ok - I'm good with that then!

Reply:
If we replaced the ZIO in your proposal with a Promise (ouch!), that should work, right?

Reply:
hmmm, no.